### PR TITLE
Centralize and simplify managed_service

### PR DESCRIPTION
**Pull request by fortuna**
This PR includes some clean up I did in the single-port branch. The idea is to start merging smaller changes rather than a huge one.

Some of the changes:

- Moved all the HTTP endpoint binding to manager_service
- Moved `ManagedAccessKeyRepository.init` logic to construction
- Removed the need for the accessKey map (it duplicates the config)
- Renamed managed_user to managed_access_key

### CODE REVIEW

File: src/shadowbox/server/managed_access_key.ts [hunk 1]
```diff
@@ -165,11 +150,11 @@ class ManagedAccessKeyRepository implements AccessKeyRepository {
       try {
         this.saveConfig();
       } catch (error) {
-        return Promise.reject(new Error(`Failed to save config: ${error}`));
+        throw new Error(`Failed to save config: ${error}`);
```
Comment (trevj) [inline src/shadowbox/server/managed_access_key.ts]: While you're in here, how about logging just `error.message`?

Also, `saveConfig` is a one liner which is only used by this function - I would just fold it in here.

File: src/shadowbox/server/managed_access_key.ts [hunk 2]
```diff
@@ -122,29 +122,14 @@ class ManagedAccessKeyRepository implements AccessKeyRepository {
...
-      startInstancePromises.push(
-          this.shadowsocksServer
-              .startInstance(
-                  accessKeyJson.port, accessKeyJson.password, this.statsSocket,
-                  accessKeyJson.encryptionMethod)
-              .then((ssInstance) => {
-                ssInstance.onInboundBytes(this.handleInboundBytes.bind(
-                    this, accessKeyJson.id, accessKeyJson.metricsId));
-              }));
+      this.startInstance(accessKeyJson);
```
Comment (trevj) [inline src/shadowbox/server/managed_access_key.ts]: Good catch removing Promise-related stuff from the constructor. This will still fail fast though, i.e. if an access key cannot be started then we won't even try to start the remainder. Is that what you want?

File: src/shadowbox/model/access_key.ts [hunk 3]
```diff
@@ -11,19 +12,24 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
-
-import {ShadowsocksInstance} from './shadowsocks_server';
-
 export type AccessKeyId = string;
 
+export interface ProxyParams {
```
Comment (trevj) [inline src/shadowbox/model/access_key.ts]: Some comments would be good, and for `AccessKey`.

File: src/shadowbox/server/managed_access_key.ts [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+  accessKeys: AccessKeyConfig[];
+  // Next AccessKeyId to use.
+  nextId: number;
+}
+
+// Generates a random password for Shadowsocks access keys.
+function generatePassword(): string {
+  return randomstring.generate(12);
+}
+class AccessKeyConfigFile {
```
Comment (trevj) [inline src/shadowbox/server/managed_access_key.ts]: blank line needed

File: src/shadowbox/server/manager_service.spec.ts [hunk 5]
```diff
@@ -30,7 +30,7 @@ describe('ShadowsocksManagerService', () => {
 
   it('lists access keys in order', (done) => {
     const repo = new MockAccessKeyRepository();
-    const service = new ShadowsocksManagerService(repo);
+    const service = new ShadowsocksManagerService(null, repo, null);
```
Comment (trevj) [inline src/shadowbox/server/manager_service.spec.ts]: Each call to the constructor in this test passes only a `repo`. DId you consider keeping that as the first argument and making the two new parameters optional?

File: src/shadowbox/model/access_key.ts [hunk 6]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (trevj) [inline src/shadowbox/model/access_key.ts]: you have a blank line here

File: src/shadowbox/model/access_key.ts [hunk 7]
```diff
@@ -11,19 +12,24 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
```
Comment (trevj) [inline src/shadowbox/model/access_key.ts]: usually we have a blank line after the copyright header

File: src/shadowbox/server/managed_access_key.ts (hunk 1 above)
Comment (fortuna) [reply src/shadowbox/server/managed_access_key.ts]: The error is logged by the caller. I think it works better if an error is only logged by whoever handles it. Otherwise you end up with inconsistent and duplicate logging.

Comment (fortuna) [reply src/shadowbox/server/managed_access_key.ts]: `saveConfig` is also called from `removeAccessKey`

File: src/shadowbox/server/managed_access_key.ts (hunk 2 above)
Comment (fortuna) [reply src/shadowbox/server/managed_access_key.ts]: Do you mean it will fail fast because the Promise rejection won't be handled? I think we only get a warning. In any case, I added error handling to the startInstance method.

File: src/shadowbox/server/manager_service.spec.ts (hunk 5 above)
Comment (fortuna) [reply src/shadowbox/server/manager_service.spec.ts]: They are actually all mandatory. I added a new test case for renames that require the server config.

File: src/shadowbox/server/managed_access_key.ts (hunk 1 above)
Comment (trevj) [reply src/shadowbox/server/managed_access_key.ts]: Sorry, I missed the other calls. For logging, I just meant change:
```ts
throw new Error(`Failed to save config: ${error}`);
```
to:
```ts
throw new Error(`Failed to save config: ${error.message}`);
```

File: src/shadowbox/server/managed_access_key.ts (hunk 2 above)
Comment (trevj) [reply src/shadowbox/server/managed_access_key.ts]: Yes, because the rejection won't be handled. The change you made to `startInstance` is better for the constructor, but it's called in other places - generally better to have the client deal with the error, no?

File: src/shadowbox/server/manager_service.spec.ts (hunk 5 above)
Comment (trevj) [reply src/shadowbox/server/manager_service.spec.ts]: More tests are good but I just meant that repeatedly constructing `ShadowsocksManagerService` with `(null, repo, null` suggests a better signature might be:
```ts
  constructor(
      private accessKeys: AccessKeyRepository,
      private serverConfig?: server_config.ServerConfig,
      private stats?: metrics.PersistentStats
  ) {}
```
Then your tests could create it with `ShadowsocksManagerService(repo)`.

Comment (fortuna) [reply src/shadowbox/server/manager_service.spec.ts]: I kind of like to make it clear that we are ignoring some of the mandatory fields, so a reader is not misled to think everything is ok. It also makes it explicit that we are not exercising the code that use those parameters, which ideally we should.

File: src/shadowbox/server/managed_access_key.ts (hunk 1 above)
Comment (trevj) [reply src/shadowbox/server/managed_access_key.ts]: Oh, this error has a proper `toString` method? I may have had my head buried in weird Windows client code for too long.

File: src/shadowbox/server/manager_service.spec.ts (hunk 5 above)
Comment (trevj) [reply src/shadowbox/server/manager_service.spec.ts]: Hmm. I think you could justify almost any signature with that reasoning. Since 90% of the calls to the constructor come from tests - not unusual - I would ordinarily think it's worth spending a little time making it elegant.
